### PR TITLE
Separate Docker image creation vs. build.

### LINKED ISSUE #49: Consider separating image creation and running build inside the image.
Currently we run the CI builds while **building** a container, in a single `docker build` command.  This downloads all the dependencies (compilers, build tools, libraries), copies the code from the CI workspace to the container, and then run the builds inside the container.

We should consider separating this in two steps.  First create a Docker image with all the dependencies, then run a container (with `docker run`) with that image that runs the build.

The advantage is that we might be able to cache the build images, that saves some time.  We can also mount the git workspace instead of using a COPY Docker command to copy the contents (around 250MiB) into the container.

The disadvantage is that when running locally the build script would leave `build-output` owned by root in your workspace.

### PR DESCRIPTION
**Pull request by coryan**
This is mostly to measure how long each phase takes, to guide any
future optimization of the builds, such as caching.  It fixes #49.

### CODE REVIEW

File: ci/build-docker.sh [hunk 1]
```diff
@@ -16,36 +16,77 @@
...
+# pointing to a (clean-ish) checkout of google-cloud-cpp:
+(cd /v ; ./ci/check-style.sh)
+
 # Run the configure / compile / test cycle inside a docker image.
 # This script is designed to work in the context created by the
 # ci/Dockerfile.* build scripts.
-mkdir -p gccpp/build-output
-cd gccpp/build-output
+readonly IMAGE="cached-${DISTRO}-${DISTRO_VERSION}"
+mkdir -p build-output/${IMAGE}
```
Comment (mbrukman) [inline ci/build-docker.sh]: "build-output/${IMAGE}"

File: ci/build-docker.sh [hunk 2]
```diff
@@ -16,36 +16,77 @@
...
+cd build-output/${IMAGE}
```
Comment (mbrukman) [inline ci/build-docker.sh]: "build-output/${IMAGE}"

File: ci/build-docker.sh [hunk 3]
```diff
@@ -16,36 +16,77 @@
...
 
 CMAKE_COMMAND="cmake"
 if [ "${SCAN_BUILD}" = "yes" ]; then
   CMAKE_COMMAND="scan-build cmake"
 fi
-${CMAKE_COMMAND} -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" ${CMAKE_FLAGS:-} ..
 
-# If scan-build is enabled we, need to manually compile the dependencies;
+echo "travis_fold:start:configure-cmake"
```
Comment (mbrukman) [inline ci/build-docker.sh]: Does Travis actually understand that as a directive and this is how it adds those collapsible sections?

File: ci/build-docker.sh [hunk 4]
```diff
@@ -16,36 +16,77 @@
...
+${CMAKE_COMMAND} -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" ${CMAKE_FLAGS:-} ../..
+echo "travis_fold:end:configure-cmake"
+
+# If scan-build is enabled, we need to manually compile the dependencies;
 # otherwise, the static analyzer finds issues in them, and there is no way to
-# ignore them.
+# ignore them.  When scan-build is not enabled, this is still useful because
+# we can fold the output in Travis and make the log more interesting.
+echo ${COLOR_YELLOW}"Started dependency build at: " $(date) ${COLOR_RESET}
```
Comment (mbrukman) [inline ci/build-docker.sh]: Quotes should include the whole thing, including vars and $(date) — they will be expanded inside the string.

File: ci/build-docker.sh [hunk 5]
```diff
@@ -16,36 +16,77 @@
...
+echo "travis_fold:start:build-dependencies"
+make -j ${NCPU} -C bigtable depends-local
+echo "travis_fold:end:build-dependencies"
+echo ${COLOR_YELLOW}"Finished dependency build at: " $(date) ${COLOR_RESET}
```
Comment (mbrukman) [inline ci/build-docker.sh]: Ditto here.

File: ci/build-docker.sh [hunk 6]
```diff
@@ -16,36 +16,77 @@
...
+
+# If scan-build is enabled we build the smallest subset of things that is
+# needed; otherwise, we pick errors from things we do not care about. With
+# scan-build disabled we compile everything, to test the build as most
+# developers will experience it.
+echo ${COLOR_YELLOW}"Started build at: " $(date) ${COLOR_RESET}
```
Comment (mbrukman) [inline ci/build-docker.sh]: And here.

File: ci/build-docker.sh [hunk 7]
```diff
@@ -16,36 +16,77 @@
...
 if [ "${SCAN_BUILD}" = "yes" ]; then
-  make -j ${NCPU} -C bigtable depends-local
   scan-build make -j ${NCPU} -C bigtable tests-local
 else
   make -j ${NCPU} all
 fi
+echo ${COLOR_YELLOW}"Finished build at: " $(date) ${COLOR_RESET}
```
Comment (mbrukman) [inline ci/build-docker.sh]: And here as well.

File: ci/build-docker.sh [hunk 8]
```diff
@@ -54,8 +95,31 @@ else
...
+then run:"
+
+DISTRO=ubuntu DISTRO_VERSION=17.04 SCAN_BUILD=yes NCPU=8 TRAVIS_OS_NAME=linux CXX=clang++ CC=clang ./ci/build-linux.sh"
+
+The HTML output will be copied into the scan-build-output subdirectory."
+${COLOR_RESET}
+_EOF_
+    exit 1
+  else
+    echo ${COLOR_GREEN}"scan-build completed without errors."${COLOR_RESET}
```
Comment (mbrukman) [inline ci/build-docker.sh]: Include the vars inside the string.

File: ci/build-docker.sh [hunk 9]
```diff
@@ -54,8 +95,31 @@ else
...
+    cp -r /tmp/scan-build-* /v/scan-build-output
+  fi
+  if [ -r scan-build-output/index.html ]; then
+    cat <<_EOF_;
+${COLOR_RED}
+scan-build detected errors.  Please read the log for details. To
+run scan-build locally and examine the HTML output install and configure Docker,
+then run:"
+
+DISTRO=ubuntu DISTRO_VERSION=17.04 SCAN_BUILD=yes NCPU=8 TRAVIS_OS_NAME=linux CXX=clang++ CC=clang ./ci/build-linux.sh"
```
Comment (mbrukman) [inline ci/build-docker.sh]: Why the quote at EOL, if you are outputting all text until `_EOF_`?

File: ci/build-docker.sh [hunk 10]
```diff
@@ -54,8 +95,31 @@ else
...
+
+The HTML output will be copied into the scan-build-output subdirectory."
```
Comment (mbrukman) [inline ci/build-docker.sh]: Why the quote at EOL, if you are outputting all text until `_EOF_`?

File: ci/build-docker.sh (hunk 1 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 2 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -123,7 +140,8 @@ which combinations are tested regularly.
  * `DISTRO`: the Linux distribution, use `ubuntu`, `fedora`, or `centos`.
  * `DISTRO_VERSION`: the version of the distribution, e.g. `17.04`.
  * `CHECK_STYLE`: if set to `yes` the build fails if the code is different
-   than the output from `clang-format(1)`.
+   than the output from `clang-format(1)`.  Notice that this reformats your
```
[the remark this answers is not in the document]
Comment (coryan) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -123,7 +140,8 @@ which combinations are tested regularly.
  * `DISTRO`: the Linux distribution, use `ubuntu`, `fedora`, or `centos`.
  * `DISTRO_VERSION`: the version of the distribution, e.g. `17.04`.
  * `CHECK_STYLE`: if set to `yes` the build fails if the code is different
```
[the remark this answers is not in the document]
Comment (coryan) [reply CONTRIBUTING.md]: Done.

File: ci/build-docker.sh (hunk 4 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 5 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 6 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 7 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 9 above)
Comment (coryan) [reply ci/build-docker.sh]: That was just a cut&paste / editing mistake.

File: ci/build-docker.sh (hunk 10 above)
Comment (coryan) [reply ci/build-docker.sh]: Fixed.

File: ci/build-docker.sh (hunk 8 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 3 above)
Comment (coryan) [reply ci/build-docker.sh]: Yes, for example see:

https://travis-ci.org/coryan/google-cloud-cpp/jobs/320346069

It is sort of documented or officially not documented in travis-ci/travis-ci#2285

**Comment by coryan** [comment]
PTAL.

**Pull request merged**